blockchain/blockdao/blob_store.go

```diff
@@ -22,4 +38,241 @@ type (
 		GetBlobsByHeight(uint64) ([]*types.BlobTxSidecar, []string, error)
 		PutBlock(*block.Block) error
 	}
+
+	// storage for past N-day's blobs, structured as blow:
+	// 1. A ring buffer is used to store these blobs. By default we store N=18
+	//    day's worth of data, meaning the size of the ring buffer is equal to
+	//    18 x 24 x 720 = 311040. The (serialized) blobs are stored using key
+	//    = 0 ~ 311039 in the bucket, and when a blob wraps around to store at
+	//    say key = 0 again, it overwrites the data 18-days ago, effectively
+	//    purging expired data.
+	// 2. Two mappings are kept in a separate index: (1) key to blob's height
+	//    and tx hashes, and (2) tx hash to the blob's height. Here's an example:
+	//    304 --> { height: 311344
+	//              []hash: eddf9e61fb9d8f5111840daef55e5fde
+	//                      95977fb340867fa13f787df864ed5388 }
+	//    and:
+	//    eddf9e61fb9d8f5111840daef55e5fde --> {height: 311344}
+	//    95977fb340867fa13f787df864ed5388 --> {height: 311344}
+	//    The 2nd mapping is used to support GetBlob(hash) function. It converts
+	//    the requested blob hash to the height, and retrieves the blob.
+	//    When a new blob overwrites an old one, it will check the 1st mapping,
+	//    and delete the hashes of old blob correspondingly.
+	// 3. The maximum number of blobs is 6 for each block, so maximum data size
+	//    stored by a key is 131kB x 6 = 786kB. The maximum total size of the
+	//    entire blob storage is 786kB x 311040 = 245GB.
+	//
+	blobStore struct {
+		kvStore                         db.KVStore
+		epochNumber, blockPerEpoch      uint32
+		totalBlocks                     uint64
+		currWriteBlock, currExpireBlock uint64
+	}
 )
+
+func NewBlobStore(kv db.KVStore, epochNum, blockPerEpoch uint32) *blobStore {
```

Comment:
prefer replace `epochNum` and `blockPerEpoch` with `size`

Reply:
changed to pass in size directly